soccer/planning/RRTPlanner.cpp

```diff
@@ -195,37 +218,380 @@ float getTime(InterpolatedPath& path, int index,
+    vector<Point> startDirections;
+    vector<Point> endDirections;
+
+    const float pathWeight = 1.0;
+    Point pathDirection = (points[1] - points[0]).normalized(pathWeight);
+    startDirections.push_back(
+        (vi + pathDirection)
+            .normalized((points[1] - points[0]).mag() * directionDistance));
+    for (int i = 1; i < curvesNum - 1; i++) {
+        const Point difference = points[i + 1] - points[i - 1];
+        endDirections.push_back(difference.normalized(
+            (points[i] - points[i - 1]).mag() * directionDistance));
+        startDirections.push_back(difference.normalized(
+            (points[i] - points[i + 1]).mag() * directionDistance));
+    }
+    Point endPathDirection = (points[points.size() - 1] -
+                              points[points.size() - 2]).normalized(pathWeight);
+    endDirections.push_back((vf + endPathDirection)
+                                .normalized((points[points.size() - 1] -
+                                             points[points.size() - 2]).mag() *
+                                            directionDistance));
+
+    vector<CubicBezierControlPoints> path;
+
+    for (int i = 0; i < curvesNum; i++) {
+        Point p0 = points[i];
+        Point p3 = points[i + 1];
+        Point p1 = p0 + startDirections[i];
+        Point p2 = p3 - endDirections[i];
+        path.emplace_back(p0, p1, p2, p3);
+    }
+    return path;
+}
 
-    if (path) {
-        return unique_ptr<Planning::InterpolatedPath>(path);
+vector<CubicBezierControlPoints> RRTPlanner::generateCubicBezierPath(
+    const vector<Geometry2d::Point>& points,
+    const MotionConstraints& motionConstraints, Geometry2d::Point vi,
+    Geometry2d::Point vf, const boost::optional<vector<float>>& times) {
+    size_t length = points.size();
+    size_t curvesNum = length - 1;
+    vector<double> pointsX(length);
+    vector<double> pointsY(length);
+    vector<double> ks(length - 1);
+    vector<double> ks2(length - 1);
+
+    for (int i = 0; i < length; i++) {
+        pointsX[i] = points[i].x;
+        pointsY[i] = points[i].y;
+    }
+    const float startSpeed = vi.mag();
+
+    const float endSpeed = vf.mag();
+
+    if (times) {
+        assert(times->size() == points.size());
+        for (int i = 0; i < curvesNum; i++) {
+            ks[i] = 1.0 / (times->at(i + 1) - times->at(i));
+            ks2[i] = ks[i] * ks[i];
+            if (std::isnan(ks[i])) {
+                debugThrow("Meh");
+                return vector<CubicBezierControlPoints>();
+            }
+        }
     } else {
+        for (int i = 0; i < curvesNum; i++) {
+            ks[i] = 1.0 / (getTime(points, i + 1, motionConstraints, startSpeed,
+                                   endSpeed) -
+                           getTime(points, i, motionConstraints, startSpeed,
+                                   endSpeed));
+            ks2[i] = ks[i] * ks[i];
+            if (std::isnan(ks[i])) {
+                debugThrow("Meh");
```

Comment:
What scenarios would lead to getting a NaN k value?  Is there something we can do to handle or prevent this better?